core/src/main/java/org/apache/druid/data/input/impl/CombiningInputSource.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.data.input.impl;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
+import org.apache.druid.data.input.AbstractInputSource;
+import org.apache.druid.data.input.InputFormat;
+import org.apache.druid.data.input.InputSource;
+import org.apache.druid.data.input.InputSplit;
+import org.apache.druid.data.input.SplitHintSpec;
+import org.apache.druid.java.util.common.Pair;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+/**
+ * InputSource that combines data from multiple inputSources. The delegate inputSources must be splittable.
+ */
+
+public class CombiningInputSource extends AbstractInputSource implements SplittableInputSource
+{
+  private final List<SplittableInputSource> delegates;
+
+  @JsonCreator
+  public CombiningInputSource(
+      @JsonProperty("delegates") List<SplittableInputSource> delegates
+  )
+  {
+    Preconditions.checkArgument(
+        delegates != null && !delegates.isEmpty(),
+        "Must specify atleast one delegate inputSource"
+    );
+    this.delegates = delegates;
+  }
+
+  @JsonProperty
+  public List<SplittableInputSource> getDelegates()
+  {
+    return delegates;
+  }
+
+  @Override
+  public Stream<InputSplit> createSplits(
+      InputFormat inputFormat,
+      @Nullable SplitHintSpec splitHintSpec
+  )
+  {
+    return delegates.stream().flatMap(inputSource -> {
+      try {
+        // Each inputSplit is paired up with its respective inputSource so that during split, withSplit() is called against
+        // the correct inputSource for each inputSplit
```

Comment:
I think this is good to document in the class-level Javadoc. Would you please add a description for overall behavior?